core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.ClassPath;
+import com.google.common.reflect.ClassPath.ClassInfo;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Sets.difference;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+import static java.lang.ClassLoader.getSystemClassLoader;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestSingleVersionBackwardCompatibility
+{
+    private static final Set<String> BACKWARD_INCOMPATIBLE_CHANGES = ImmutableSet.of(
+            "Constructor: public io.trino.spi.connector.TableProcedureExecutionMode(boolean,boolean)",
+            "Constructor: public io.trino.spi.eventlistener.QueryStatistics(java.time.Duration,java.time.Duration,java.time.Duration,java.time.Duration,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,long,long,long,long,long,long,long,long,long,long,long,long,long,long,long,double,double,java.util.List<io.trino.spi.eventlistener.StageGcStatistics>,int,boolean,java.util.List<io.trino.spi.eventlistener.StageCpuDistribution>,java.util.List<java.lang.String>,java.util.Optional<java.lang.String>)");
+
+    @Test
+    public void testSpiSingleVersionBackwardCompatibility()
+            throws Exception
+    {
+        assertThat(getCurrentSpi())
+                .containsAll(difference(getPreviousSpi(), BACKWARD_INCOMPATIBLE_CHANGES));
```

Comment:
exclusions (BACKWARD_INCOMPATIBLE_CHANGES) should be verified for being part of previous SPI

eg a test should fail when i add `public void foo()` to BACKWARD_INCOMPATIBLE_CHANGES, unless such method existed in the previous SPI AND it no longer exists in the new SPI.

Reply:
Alternatively, we could eg have `BACKWARD_INCOMPATIBLE_CHANGES` as a map `version -> changes`
and we would just look at the entry for current version.
Of course, the point is not to keep track of all the breaking changes ever, but rather not to fail right after a release.
Then next person adding something new to the list would remove old (obsolete) entries.

Reply:
I no longer get a version from the pom to java. So in tests I have no notion of a version. Bringing back the version in the test just for this sounds like an overkill to me.

Maybe I can simply add a comment that:

```
// when updating  BACKWARD_INCOMPATIBLE_CHANGES please check if you can remove obsolete entries
```
?

Reply:
> Bringing back the version in the test just for this sounds like an overkill to me.

It's not

BTW i am making Java aware of SPI version in https://github.com/trinodb/trino/pull/11774